statistics/handle/statscache.go

```diff
@@ -0,0 +1,204 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package handle
+
+import (
+	"github.com/pingcap/tidb/statistics"
+)
+
+// statsCacheInner is the interface to manage the statsCache, it can be implemented by map, lru cache or other structures.
+type statsCacheInner interface {
+	Get(int64) (*statistics.Table, bool)
+	Put(int64, *statistics.Table)
+	Del(int64)
+	Cost() int64
+	Keys() []int64
+	Values() []*statistics.Table
+	Map() map[int64]*statistics.Table
+	Len() int
+	FreshMemUsage()
+	FreshTableCost(int64)
+	Copy() statsCacheInner
+}
+
+type cacheItem struct {
+	key   int64
+	value *statistics.Table
+	cost  int64
+}
+
+func newStatsCache() statsCache {
+	return statsCache{
+		statsCacheInner: &mapCache{
+			tables:   make(map[int64]cacheItem),
+			memUsage: 0,
+		},
+	}
+}
+
+// statsCache caches the tables in memory for Handle.
+type statsCache struct {
+	// version is the latest version of cache. It is bumped when new records of `mysql.stats_meta` are loaded into cache.
+	version uint64
+	// minorVersion is to differentiate the cache when the version is unchanged while the cache contents are
+	// modified indeed. This can happen when we load extra column histograms into cache, or when we modify the cache with
+	// statistics feedbacks, etc. We cannot bump the version then because no new changes of `mysql.stats_meta` are loaded,
+	// while the override of statsCache is in a copy-on-write way, to make sure the statsCache is unchanged by others during the
+	// the interval of 'copy' and 'write', every 'write' should bump / check this minorVersion if the version keeps
+	// unchanged.
+	// This bump / check logic is encapsulated in `statsCache.update` and `updateStatsCache`, callers don't need to care
+	// about this minorVersion actually.
+	minorVersion uint64
+
+	statsCacheInner
```

Comment:
Are you going to maintain LRU info for column stats in this structure?

Reply:
right